src/xdsl/error_format.py

```diff
@@ -0,0 +1,103 @@
+import os
+import sys
+from types import TracebackType
+
+
+class Colors:
+
+    class format:
+        reset = '\033[0m'
+        bold = '\033[01m'
+        underline = '\033[04m'
+
+    class fg:
+        red = '\033[31m'
+        green = '\033[32m'
+        orange = '\033[33m'
+        blue = '\033[34m'
+        cyan = '\033[36m'
+
+
+class Frame:
+    """
+    This class is a traceback wrapper of a given raised error 
+    """
+    e: Exception
+    extra_lines: int
+
+    def __init__(self, e: Exception, extra_lines: int):
+        self.e = e
+        self.extra_lines = extra_lines
+
+    def get_frame(self, num: int):
+        """
+        Extracting the frame from the traceback.
+        With given exception
+        """
+        _, _, exc_traceback = sys.exc_info()
+
+        stack: list[TracebackType] = []
+        tb = exc_traceback
+        while tb:
+            stack.append(tb)
+            tb = tb.tb_next
+        stack = stack[-num:] if num else []
+
+        for tb in stack:
+            frame = tb.tb_frame
+            code = frame.f_code
+
+            filename = code.co_filename
+            line_num = frame.f_lineno
+            exc_name = code.co_name
+            local_var = frame.f_locals
+
+            yield filename, line_num, exc_name, local_var, self.extract_code(
+                filename, line_num)
+
+    def extract_code(self, filename: str, line_num: int):
```

Comment:
That is indeed doable, might take a bit of time with regex (as always)

Reply:
Yes, the first line of the function is indicated by ```code.co_firstlineno```.
I don't have ideas for the last line of function yet. Maybe we can use with regex as we have the current function name in ```exc_name``` and the compiled bytecode of the function in ```code.co_code```.

Side note:
We can do some hack on the compiled bytecode ```code.co_code``` to get a more accurate position of where the exception is called. But it may be tedious work.